### PR TITLE
llvm: set PIE only for executables

### LINKED ISSUE #17575: Android link error: relocation R_ARM_TLS_LE32 against tl.foo cannot be used with -shared
### Zig Version

0.12.0-dev.983+78f2ae7f2

### Steps to Reproduce and Observed Behavior

Take this program:

```zig
// tl.zig
const std = @import("std");

threadlocal var foo: std.Thread.Id = 0;

export fn setFoo() void {
    foo = std.Thread.getCurrentId();
}

export fn getFoo() std.Thread.Id {
    return foo;
}
```

We can build a shared lib for linux with the following command:

```
zig.exe build-lib -target arm-linux -dynamic -fPIC tl.zig
```

The previous command compiles fine.

Now if we change `arm-linux` to `arm-linux-android`:

```
zig.exe build-lib -target arm-linux-android -dynamic -fPIC tl.zig
```

We get the following compile error:

```
error: ld.lld: relocation R_ARM_TLS_LE32 against tl.foo cannot be used with -shared
    note: defined in libtl.so.o
    note: referenced by tl.zig:0
    note:               libtl.so.o:(setFoo)
error: ld.lld: relocation R_ARM_TLS_LE32 against tl.foo cannot be used with -shared
    note: defined in libtl.so.o
    note: referenced by tl.zig:0
    note:               libtl.so.o:(getFoo)
error: ld.lld: relocation R_ARM_TLS_LE32 against Thread.LinuxThreadImpl.tls_thread_id cannot be used with -shared
    note: defined in libtl.so.o
    note: referenced by Thread.zig:0 (C:\Users\tuket\Documents\APPS\zig_experimental\lib\std/Thread.zig:0)
    note:               libtl.so.o:(Thread.LinuxThreadImpl.getCurrentId)
error: ld.lld: relocation R_ARM_TLS_LE32 against debug.panic_stage cannot be used with -shared
    note: defined in libtl.so.o
    note: referenced by debug.zig:0 (C:\Users\tuket\Documents\APPS\zig_experimental\lib\std/debug.zig:0)
    note:               libtl.so.o:(debug.panicImpl)
```

However, if we run the same command with zig 0.11.0, it compiles fine.

### Expected Behavior

I'm not certain what should be the expected behavior, but it's weird that it used to work and now it doesn't.

This is preventing me from compiling Android apps.

For this bud report, I tried to make the program as simple as possible.

In my more elaborate Android program, I wasn't actually using the `std.Thread.getCurrentId()` function. In fact, when I would add the "-Drelease" flag to zig, the compiler error would go away (probably due to unused code elimination).

### PR DESCRIPTION
**Pull request by mikdusan**
closes #17575

### CODE REVIEW

**Comment by andrewrk** [comment]
Shouldn't `options.pie` be set to false earlier in the frontend?

**Review by andrewrk — CHANGES_REQUESTED** [review]
missing error when the user specifies `-fpie` for an executable

**Comment by mikdusan** [comment]
> missing error when the user specifies `-fpie` for a non-executable

ha I actually had that done and took it out.

Do we also want to error for `-fno-PIE` for non-executables?

**Comment by andrewrk** [comment]
no

File: src/Compilation.zig [hunk 1]
```diff
@@ -1037,13 +1037,18 @@ pub fn create(gpa: Allocator, options: InitOptions) !*Compilation {
...
-                return error.TargetRequiresPIE;
+        const pie: bool = pie: {
+            if (options.output_mode == .Exe) {
+                const must_pie = target_util.requiresPIE(options.target);
+                if (options.want_pie) |explicit| {
+                    if (!explicit and must_pie) return error.TargetRequiresPIE;
+                    break :pie explicit;
+                } else break :pie must_pie or tsan;
+            } else {
+                if (options.want_pie) |explicit| if (explicit) return error.PIERequiresExe;
```
Comment (andrewrk) [inline src/Compilation.zig:1048]:
```suggestion
                if (options.want_pie == true) return error.PIERequiresExe;
```

**Review by andrewrk — CHANGES_REQUESTED** [review]
keep it simple:

```zig
const pie: bool = pie: {
    if (options.output_mode != .Exe) {
        if (options.want_pie == true) return error.OutputModeForbidsPie;
        break :pie false;
    }
    if (target_util.requiresPIE(options.target)) {
        if (options.want_pie == false) return error.TargetRequiresPie;
        break :pie true;
    }
    if (tsan) {
        if (options.want_pie == false) return error.TsanRequiresPie;
        break :pie true;
    }
    if (options.want_pie) |want_pie| {
        break :pie want_pie;
    }
    break :pie false;
};
```

**Comment by andrewrk** [comment]
Looks good, thanks!

**Comment by IntegratedQuantum** [comment]
This change appears to break the thread sanitizer even further:
I was just attempting to look into repairing the thread sanitizer, when I got
`error: unable to build TSAN library: OutputModeForbidsPie`
It appears that the thread sanitizer builds as a lib, but also requires PIE.

**Comment by mikdusan** [comment]
> This change appears to break the thread sanitizer even further: I was just attempting to look into repairing the thread sanitizer, when I got `error: unable to build TSAN library: OutputModeForbidsPie` It appears that the thread sanitizer builds as a lib, but also requires PIE.

#17653 should fix it

**Pull request merged**